### PR TITLE
Adding Manipulating Hierarchies in Atoti Notebook

### PR DESCRIPTION
**Pull request by andrewyang94**
This PR adds a notebook on using the following Hierarchy class properties and functions with storytelling elements for multi-level hierarchies, slicing hierarchies, and virtual hierarchies.

* Basic
   * `dimension`
   * `dimension_default`
   * `levels`
   * `name`
   * `visible`
* Advanced
   * `slicing`
   * `virtual`
   * `isin()`

> **Note:** This notebook is designed to be a follow-up to https://github.com/atoti/atoti/pull/877, and therefore some content from the previous notebook (Hierarchy Fundamentals in Atoti) has been moved to this notebook.

### CODE REVIEW

File: 02-technical-guides/hierarchies-advanced/main.ipynb [hunk 1]
```diff
@@ -0,0 +1,1998 @@
+{
```
Comment (andrewyang94) [inline 02-technical-guides/hierarchies-advanced/main.ipynb:1]: <p>@tibdex I noticed that when I set my <code>OrderId</code> hierarchy as a virtual hierarchy, I notice and <strong>increase</strong> in memory footprint rather than a <strong>decrease</strong>. Perhaps this is just specific to my environment? 🤷‍♂️ When you get a chance, could you try to run through this section of the notebook and reproduce the issue? Not sure why this would be the case.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='62'/>

**Comment by andrewyang94** [comment]
@tibdex When you have a chance, can you also verify my rationale/technical interpretation of the advanced props + functions that I noted? Namely:

* `Advanced`
    * `slicing`
    * `virtual`
    * `isin()`

File: 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb [hunk 2]
```diff
@@ -0,0 +1,2029 @@
+{
```
Comment (CateF22) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>Add in a bit here:</p><p> </p><p> </p><h2>Basic Hierarchy Properties </h2><p>Each hierarchy is an object of the <a href="https://docs.atoti.io/latest/api/atoti.Hierarchy.html#atoti.Hierarchy" target="_blank">atoti.Hierarchy</a> class. This means that we are able to interact with and manipulate hierarchies using their associated class properties and functions. </p><p><strong>Some of the properties allow us to understand how the hierarchies in the cube are structuresd, and this will really help us make more sensible decisions when it comes to data analysis. (this is the new bit)</strong></p><p>A few basic properties are as follows: </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (CateF22) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>Add in a bit at the top to quickly explain what a parent hierarchy is.</p><p>Otherwise I might be a bit lost, and not understand why this is interesting for me.</p><p> </p><p>A quick suggestion:</p><p>A hierarchy is by definition a set of parent-child relationships, and in Atoti we say that a hierarchy is the parent of another when the second hierarchy is a lower level.</p><p>In our hierarchies Products is the parent to Product Category, and so Product Category is the child of Products.</p><p>It is useful to know how hierarchies are structured, as this will help us decide which level of a hierarchy we should analyze, and how far up or down the hierarchy we should aggregate the data to get meaningful insights.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (CateF22) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>I don't see how knowing which one is the default hierarchy helps me.</p><p>Explain somehow what this features means and what added value it brings to the user</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (CateF22) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>this should probably be the first property mentioned, as knowing which levels you have helps to understand and explain the parent-child relationship</p><p>Move this to before the dimension property</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (CateF22) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>why would people need to do this?</p><p>can you give some context to when this would be a beneficial use of time?</p><p>Every decision is a cost-benefit analysis, so before I endeavor to changes something (cost) I need to know what I will be able to do in the future that I can't do today (benefit)</p><p>Add some context into this section (or delete it?)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Comment (CateF22) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>Add in a sentence to say why it's useful to have multi-level hierarchies and not lots of single level ones</p><p>You could mention the UI and the you'll be able to drill into the data or aggregate to only a given level</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

Comment (CateF22) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>Change this: </p><p> </p><p>Slicing hierarchies are used when we want a hierarchy to no longer be aggregable at the top level/across all members of a hierarchy. </p><p> </p><p>To =&gt;</p><p>The values associated with a hierarchy are by default aggregated to the top level, across all members of that hierarchy.  This makes sense and provides useful analysis for data that should be summed in this way.  But for some hierarchies, it makes more sense to specify one of the hierarchy members as a default or slicing member.  This is very common for dates, so that today's date is displayed instead of the aggregate of all the values since the beginning of time</p><p>(you'll have to reword that and paraphrase it a bit)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='51'/>

Comment (CateF22) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>Is this related to slicing hierarchies?</p><p>I suspect it might be in the wrong spot</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='54'/>

Comment (CateF22) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>this bit about slicing hierarchies is confusing for me, as I'm used to seeing them used for dates, and not for other things.</p><p>Check with Hetal and see what she suggests doing.</p><p>There may be real use cases when you would use a slicing hierarchy for something that is not a date, but I can't see why or when </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='59'/>

Comment (CateF22) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>from here on, I'm a bit confused</p><p>and I'm not sure if there was a problem uploading an image or if the text I can see is what I should be seeing</p><p>If this is the correct text, it would be wise to contextualize and explain it</p><p>Maybe also for an image</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='67'/>

Comment (CateF22) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>same comment about the image, I'm not sure if I can see the thing that I am supposed to see,</p><p>it's a bit confusing</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='70'/>

Comment (printhellohetal) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>Minor amendment, please, to keep it all in active voice, as per the brand guidelines :)</p><p> </p><p>Some of the properties allow us to understand how the structure of the hierarchies in the cube works, and this will really help us make more sensible decisions when it comes to data analysis.</p><p> </p><p>or</p><p> </p><p>Some of the properties allow us to understand the structure of the hierarchies in the cube, and this will really help us make more sensible decisions when it comes to data analysis.</p>

Comment (printhellohetal) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>This explanation is good, but I think it is missing a bit about dimensions. If incorporating this bit, let's make sure the relationship between dimensions and hierarchies are clear (and it might require taking a step back to explain what is a dimension).</p>

Comment (printhellohetal) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>This belongs here, but how this relates to slicing hierarchies is getting a bit lost in the weeds of <code>create_parameter_hierarchy</code> and <code>at</code>  and such.</p><p> </p><p>It would be worth seeing how to make this and the above more streamlined (the <code>filter</code>  and <code>isin</code>  is a bit advanced if the person hasn't learned about measures yet...)   </p>

Comment (printhellohetal) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>Dates is a more canonical topic to use for explaining slicing hierarchies for sure.</p><p> </p><p>We don't need to lose this bit, but it might be worth separating the two sections out: slicing hierarchies using a simple example, then <code>create_parameter_hierarchy_from_column</code> , then mention it creates a slicing hierarchy by default "for free"</p>

File: 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb [hunk 3]
```diff
@@ -0,0 +1,1592 @@
+{
```
Comment (printhellohetal) [inline 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb:1]: <p>Minor point here: I wouldn't classify Order ID an 'inherently numerical column' as one can choose an ID convention that includes letters. Columns that are inherently numerical are things like "year" or "price." And here, the first would be better as a hierarchy.  I think this can be addressed by simply removing the word "inherently."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (printhellohetal) [inline 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb:1]: <p>Same comment as on the other notebook, for consistency's sake, can we use "alias" instead of "set variables" language in this comment</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (printhellohetal) [inline 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb:1]: <blockquote>however, for the purposes of this walk through, we've elected to use <code>session.widget</code> so that we can show our analysis directly within this Jupyter notebook.</blockquote><p>No need to change this here, just food for thought:</p><p>Yes, we use the extension for creating a story in a Jupyter notebook, but the extension also provides value to other users, allowing them to do iterative analysis, et cetera. We should be careful when we talk about it that we also highlight how this extension exists as a tool to help them.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

File: 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb [hunk 4]
```diff
@@ -0,0 +1,2036 @@
...
```
Comment (printhellohetal) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>To be consistent, can we change the comment here to specifically say "alias" instead of the "set variables" language.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (printhellohetal) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>Instead of "we can manipulate" it might be worth swapping to something like "we will need to manipulate" or "we should manipulate"</p><p>The current sentence suggests the below cell will somehow supply us with a note or tag that says whether or not it can be manipulated, whereas really, users would do this to check to see if they have any work to do.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (printhellohetal) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>I'm not sure what the phrase "fully qualified list of levels" means here. Is this just to say "the full list of levels"</p><p> </p><p>Edit: I realize you mean "fully qualified" in the technical sense, but I think you only briefly define it above. It might deserve a bit more robustness if you're going to use it quite a bit. At which point it would suggest explaining why someone would want the fully qualified list.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (printhellohetal) [inline 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>A screenshot with boxes would be better.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='67'/>

File: 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb (hunk 2 above)
Comment (andrewyang94) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>Hmm not sure if I like the wording of this:</p><p> </p><p><strong>"Some of the properties allow us to understand how the structure of the hierarchies in the cube works..."</strong></p><p> </p><p>It makes it seem as though knowing the properties is a dependency for understanding the structure of the hierarchies in the cube. I've amended this to the following as a reword:</p><p> </p><p><strong>"Using hierarchy properties give us additional insight into the specifics of each hierarchy in the cube, and help us make more sensible decisions when it comes to data analysis."</strong></p><p> </p><p>I feel like the amended version captured the intended connotation of the proposed amendment, but let me know otherwise!</p>

Comment (andrewyang94) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>I'm not sure we need to discuss parent/multi-level hierarchies in this passage, as I'm just talking about hierarchies and the dimensions that they fall under. Therefore, I've amended this to:</p><p> </p><p>"The&nbsp;<code>dimension</code>&nbsp;property of each hierarchy returns the name of its parent dimension. This is helpful for when we want to know which dimension a particular hierarchy falls within. For example, we can call the&nbsp;<code>dimension</code>&nbsp;property of the&nbsp;<code>Address</code>&nbsp;hierarchy to determine that its parent dimension is&nbsp;<code>Customers</code>.</p><blockquote>💡&nbsp;<strong>Note:</strong>&nbsp;Remember that when we loaded our tables into our cube, the tables effectively became our dimensions and the non-numeric columns became our hierarchies. Therefore, the&nbsp;<code>Address</code>&nbsp;column of our&nbsp;<code>Customer</code>&nbsp;table became the&nbsp;<code>Address</code>&nbsp;hierarchy under the&nbsp;<code>Customer</code>&nbsp;dimension of our cube, respectively. See the previous cube data model/schema visualization for reference."</blockquote><p> </p>

Comment (andrewyang94) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>Mmm good point. I amended this by adding the description in our docs for <a href="https://docs.atoti.io/latest/api/atoti.Hierarchy.dimension_default.html#atoti.Hierarchy.dimension_default" target="_blank">https://docs.atoti.io/latest/api/atoti.Hierarchy.dimension_default.html#atoti.Hierarchy.dimension_default</a>. Amended statement is now:</p><p> </p><p>"<strong>Every dimensions in the cube has a designated default hierarchy. This is helpful as some UIs support clicking on a dimension (or dragging and dropping it) as a shortcut to add its default hierarchy to a widget.</strong></p><p> </p><p><strong>Calling the&nbsp;</strong><a href="https://docs.atoti.io/latest/api/atoti.Hierarchy.dimension_default.html#atoti.Hierarchy.dimension_default" target="_blank"><strong>dimension_default</strong></a><strong>&nbsp;property of a hierarchy returns&nbsp;</strong><code><strong>True</strong></code><strong>&nbsp;or&nbsp;</strong><code><strong>False</strong></code><strong>&nbsp;for whether the specified hierarchy is the default hierarchy of its parent dimension. For example, if we call of the&nbsp;</strong><code><strong>dimention_default</strong></code><strong>&nbsp;property of our&nbsp;</strong><code><strong>Address</strong></code><strong>&nbsp;hierarchy, we see the following:</strong>"</p><p> </p><p>and also amended a later statement down the line:</p><p> </p><p>"<strong>A new default hierarchy for a dimension can be assigned as well. For example, we can set our&nbsp;</strong><code><strong>Address</strong></code><strong>&nbsp;hierarchy as the default hierarchy for the&nbsp;</strong><code><strong>Customers</strong></code><strong>&nbsp;dimension like so:</strong>"</p><p> </p><p> </p>

Comment (andrewyang94) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>I added the following explanation:</p><p> </p><p>"using the&nbsp;<code>name</code>&nbsp;property may be more helpful when trying to execute a&nbsp;<code>for</code>&nbsp;loop to iterate over all hierarchies in an Atoti data model, where we may want the colloquial hierarchy name rather than the fully qualified one (with both the dimension and hierarchy listed)."</p><p> </p><p>But I inferred this based on what I see, i.e. something like <code>h['Address'].name</code> seems super redundant to me so I can only imagine that the property comes to play when iterating over hierarchies similar to what I explained. If there is a more specific use case or description for <em>why</em> we need this property, I think someone like @tibdex or another R&amp;D person would have to chime in. </p>

Comment (andrewyang94) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>After speaking with @printhellohetal, I've re-designed the <code>Using slicing hierarchies</code> section in general. This now reads as the following, with a corresponding code walkthrough:</p><p> </p><p><strong>Slicing hierarchies are used when we want a hierarchy that is no longer aggregable at the top level/across all of its underlying members. For example, within our data model imagine that for every order made, a corresponding record of inventory with an&nbsp;</strong><code>AsOfDate</code><strong>&nbsp;is taken for each product sold. We can assume that each product started at&nbsp;</strong><code>10,000</code><strong>&nbsp;items in inventory. In this case, it would not provide any meaningful information to aggregate the current state or quantity of inventory across all the&nbsp;</strong><code>AsOfDate</code><strong>&nbsp;member dates, nor would it be a good use of computational resources.</strong></p><p> </p><p><strong>Let's see how this works in practice.</strong></p>

Comment (andrewyang94) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>I removed this section as introducing <code>create_parameter_hierarchy_from_column</code> seems a bit too complex at this point in time of the hierarchies series. We may re-introduce this content in the next part 3 hierarchies notebook that will refer to source simulations and parameter hierarchies (is what I'm thinking as of now, but this may change) </p>

Comment (andrewyang94) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>The <code>Using slicing hierarchies</code> section has been re-designed to include a more basic example with <code>AsOfDate</code> and inventory amounts. The content for <code>create_parameter_hierarchy_from_column</code> has been taken out and will most likely be moved to the 3rd part notebook for hierarchies, which will refer to parameter hierarchies and source simulations.</p>

Comment (andrewyang94) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>Replaced this with an image!</p>

Comment (andrewyang94) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>Replaced this with an image, along with a note callout that further explains the difference between direct vs. heap memory in Java.</p>

File: 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb (hunk 3 above)
Comment (andrewyang94) [reply 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb:1]: <p>Mmmm very good point! Took out the word "inherently".</p>

Comment (andrewyang94) [reply 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb:1]: <p>This is fixed, thanks for catching this!</p>

Comment (andrewyang94) [reply 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb:1]: <p>Good point, I changed this to the following to accommodate what you mentioned:</p><p> </p><p><strong>💡&nbsp;Note: You may also use&nbsp;</strong><code><strong>session.link</strong></code><strong>&nbsp;to perform the same investigation within the Atoti UI; however, for the purposes of this walk through, we've elected to use&nbsp;</strong><code><strong>session.widget</strong></code><strong>&nbsp;so that we can show our analysis directly within this Jupyter notebook and provide storytelling as we create several visualizations in Atoti.</strong></p>

File: 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb (hunk 4 above)
Comment (andrewyang94) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>This is done, thanks for catching this!</p>

Comment (andrewyang94) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>Great connotative point, I changed this to the following:</p><p> </p><p><strong>Great, looks like we're set up. Let's now take a look at the list of our hierarchies to see which ones we will look to manipulate.</strong></p>

Comment (andrewyang94) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>This is a great point, when I was mentioning the "fully qualified list of levels" in the <code>.levels</code> section, it was not with the same intent as when I was talking about the "fully qualified one (with both the dimension and hierarchy listed)" for the <code>.name</code> section. Therefore, I've amended the <code>.levels</code> section to be the following:</p><p> </p><p><strong>But let's say we wanted to see just the levels for our previous multi-level hierarchy&nbsp;</strong><code><strong>Product</strong></code><strong>. Scanning through the full list of levels in the cube may be a bit cumbersome (especially if we have a large amount of levels!). Thus, we can instead use the&nbsp;</strong><code><strong>levels</strong></code><strong>&nbsp;property for the multi-level&nbsp;</strong><code><strong>Product</strong></code><strong>&nbsp;hierarchy object to only show the levels for that particular hierarchy.</strong></p>

Comment (andrewyang94) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb:1]: <p>Done!</p>

File: 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb (hunk 2 above)
Comment (printhellohetal) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb]: Can we nix the "Using" in your counter-proposal?

File: 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb (hunk 3 above)
Comment (printhellohetal) [reply 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb:1]: Let's stick to "we" instead of "you" in our writing. The usage of "you" should be rare, and only if necessary.

File: 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb (hunk 2 above)
Comment (andrewyang94) [reply 02-technical-guides/hierarchies/02-hierarchy-manipulations.ipynb]: Fixed!

File: 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb (hunk 3 above)
Comment (andrewyang94) [reply 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb:1]: Fixed!

File: 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb [hunk 5]
```diff
@@ -0,0 +1,1605 @@
...
```
Comment (printhellohetal) [inline 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb:1]: <p>if <strong><em>we</em></strong> create <strong><em>our</em></strong> cube in auto or no_measures mode... </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (andrewyang94) [reply 02-technical-guides/hierarchies/01-hierarchy-fundamentals.ipynb:1]: <p>Fixed!</p>

**Comment by andrewyang94** [comment]
<p>Re-worded this to the following:</p><p> </p><p><strong>💡&nbsp;Note:&nbsp;You may have noticed that we have&nbsp;</strong><code><strong>Totals</strong></code><strong>&nbsp;appear in our pivot table despite us calculating single value measures at the lowest granularity of our inventory and order records. This is because the&nbsp;</strong><a href="https://docs.atoti.io/latest/api/atoti.agg.single_value.html#atoti.agg.single_value" target="_blank"><strong>single_value</strong></a><strong>&nbsp;function considers that "<em>if the value is the same for all members of the level the operand is being aggregated on, it will be propagated to the next level.</em>"</strong></p><p> </p><p>As discussed, this may be one of those particular instances where "you" does make sense to say, since I'm highlighting something that the user may have noticed, and addressing the reasoning behind what it is that they may have seen. Though open to re-wording this further!</p>

---

 <a href='https://app.reviewnb.com/atoti/atoti/pull/890/discussion/#comment-2176910699'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Review by printhellohetal — APPROVED** [review]
Think these notebooks are good to go.

Bonus if we get a readme, but I won't hold up the review process for it :)

**Comment by andrewyang94** [comment]
> Think these notebooks are good to go.
> 
> Bonus if we get a readme, but I won't hold up the review process for it :)

@printhellohetal Great point, for some reason the technical guides README also got removed so I added this back in. In addition, I created a short README for the hierarchies directory to explain what the two notebooks do.

It should also be noted that the current set up for generating our READMEs does not currently account for multiple notebooks in a single directory. This will need to be something that we account for when we refactor the README generator in the future.

image
image
